doc/developer/design/20240522_mir_representation_types.md

```diff
@@ -0,0 +1,133 @@
+# MIR Representation Types
+
+- Associated:
+  + https://github.com/MaterializeInc/materialize/issues/4171
+  + https://github.com/MaterializeInc/materialize/pull/27029
+  + https://github.com/MaterializeInc/materialize/pull/27109
+
+MIR is typechecked using user-facing SQL types.
+SQL types make distinctions between types (e.g., `CHAR(5)`, `VARCHAR(5)`, and `TEXT`) that will not be realized at runtime (all of these are represented using `Datum::String`).
+
+This design doc proposes _representation types_---a new notion of type for MIR terms.
+A representation type maps to a subset of constructors/variants of the `Datum` enum.
+
+Making this change should help address issues in optimization, particularly index selection (see ["The Problem"](#the-problem)).
+It will also simplify and clarify what MIR is doing.
+
+## The Problem
+
+[Postgres has several types for character data:](https://www.postgresql.org/docs/current/datatype-character.html)
+
+- `char(n)`, `bpchar(n)` for fixed-length strings with a fixed-length representation (whitespace padded)
+- `varchar(n)` for fixed-length strings with a variable-length representation
+- `varchar` for arbitrary-length strings with a variable-length representation
+- `bpchar` for artbirary-length strings with a variable-length representation where trailing whitespace is trimmed
+- `text` for arbitrary-length strings with a variable-length representation
+
+(There are also `"char"` and `name`, but they are very much internal to Postgres---let's ignore them for the moment.)
+
+Materialize implements all five of these character sequence types using `Datum::String`.
+Materialize only implements meaningful operations on the `text` type; to operate on other character sequence types, one must convert to `text`.
+Conversion to `text` is a no-op at runtime, since we already have the string.
+We freely insert implicit coercions to `text` in the early stages of query processing.
+For correct error behavior, however, we need to carefully keep track of failures to meet length limits.
+
+These implicit coercions---often of the form `varchar_to_text(#c)` for some column number `c`---get in the way during join planning.
+Even if we have an index on column `c` as a `varchar`, we cannot use that index to plan a join on `varchar_to_text(#c)`: not only are the types different, the terms are different, too.
+We know these implicit coercions interfere with joins; they may also interfere with other aspects of optimization.
```

Comment:
There are two related ideas here:

 1. what types characterize integers?
 2. how do we deal with casts and their possibly-erroring inverses?

## What types characterize integers?

My primary goal was to have the types at the MIR level characterize the `Datum`s the operator would deal with. All of the various string types will just be `Datum::String` at runtime, so one type works for them. But we actually have different numeric `Datum` constructors: many different sizes of ints (in unsigned and signed flavors),  two sizes of floats, and arbitrary precision numerics. So If MIR types characterize `Datum` constructors, then it seems we would probably need different numeric types.

I can imagine a system where there's a unified numeric supertype `Num` which subsumes all of these individual types, i.e., a value of type num could be any possible numeric `Datum`. But a secondary goal is to avoid complex/automatic typing relationships; one proxy for this complexity is how much overlap there is in the mapping from MIR types to `Datum` constructors. It's unfortunately not possible to have an injective mapping from MIR types to `Datum` constructors, because `Datum::True` and `Datum::False` are booleans but also JSON values... but adding a `Num` supertype would be fairly complicating. And, it wouldn't resolve our other issue...

## How do we deal with casts and their possibly-erroring inverses?

Most of our conversions have inverses---`text_to_varchar` has `varchar_to_text`, timestamps can be converted to and from strings, and the numeric types can move into and out of each other. But in general, these are not true inverses/bijections: one side might produce an error.

⚠️ **Unnecessary math jargon**: ⚠️ `text_to_varchar` is a left inverse of `varchar_to_text`, but not a right inverse. That is, `text_to_varchar[k](varchar_to_text[k](s)) == s` for all `s` of type `varchar(k)`, for all `k`. But it is not the case that `varchar_to_text[k](text_to_varchar[k](s))` for all `s` of type `text` for all `k`---it could be that `s` is longer than `k` and `text_to_varchar[k](s)` produces an error. Put differently, `varchar_to_text` is a _section_ and `text_to_varchar` is a _retraction_.

In general, there's no way to remove the fallible casts and adding in a catch-all numeric type without introducing something worse: pervasive checks and coercions between numeric types.

I think the right way forward here is:

1. Use `UnaryFunc::could_error()` with `UnaryFunc::inverse()`---if neither can error, we can simplify `f(f_inv(x))` to `x`.
2. Identify left-inverses explicitly: if we have `f(f_inv(x))` and only the outer one is fallible, we have a left inverse and we can simplify it to `x`.
3. Explore optimizations around numeric sizing and index selection.

I'm starting to see your point about index selection being a key moment; I'll address it in my response to your other comment.

Reply:
When you say "Identify left-inverses explicitly", you mean that we'd add a new function property, i.e., it's not enough to simply call `could_error` on the inner one, right? Because, in theory, the outer one might fail for some other reason besides not being able to represent the value in the original type. E.g., one might imagine a (thoroughly contrived) function `varchar_to_text_if_not_empty`, which errors if the varchar is empty. This function would be an `inverse` of `text_to_varchar` in the sense that the current code is using the word `inverse`, which is something like "if both `preserves_uniqueness` then it's an inverse for such values for which neither errors".

But maybe we could redefine our `inverse` to mean that you just have to check `could_error` on the inner one. Probably the above contrived case is not actually occurring for any of our functions.

Btw. I just found [an old Slack thread](https://materializeinc.slack.com/archives/C0246GEHL8N/p1692997054843739?thread_ts=1692994729.626079&cid=C0246GEHL8N) where I wrote more details on what the current code means by `inverse` than what to code comment is saying. (I had it on my todo list to actually improve the code comment with this information, but somehow it got buried.)

Reply:
You're absolutely right that if we had a function with a declared inverse that errored for some other reason we would have a problem with that approach.

I agree that we should rethink/refine how we think about inverses. Having a single function that says "this is my inverse and here's whether or not I could error" would simplify the thinking here.